### ISSUE TITLE
fine grain access control for groups of users and groups of repositories within a harbor project

### ISSUE THREAD

**Issue by xaleeks** [goharbor/harbor#10159]
Currently, harbor admin can create harbor projects (highest hierarchy) and users added to a project get access to all repositories created within it. 

Changes needed for finer grain access controls
- Allow users to be configured to have access to a certain set of repositories within a harbor project (still call this ‘project’ for now).  Access to repositories can be added or removed at any time.
- Allow robot accounts to be configured to have access to a certain set of repositories within a harbor project. Access to repositories can be added or removed at any time.  
- Allow grouping of users & robot accounts into groups that can then be mapped to specific repos for access, so it’s not all or nothing.  Members can be added or removed from these at any time.
- Allow configurable permissions (read, write, admin) at the repository level so users or groups of users can get different permissions for different repositories
- Needs to resolve with current system of role-based access control where guest / dev / master / admin have specific set of permissions

**Comment by xaleeks** [comment]
https://docs.google.com/document/d/1AM9XEUjSepv7EgvjSpJXP-2pSBTrExoJsj8XKYN1sZ4/edit

discussions here, ping me for access

**Comment by turnham** [comment]
My team is very interested in this feature.  I'm not sure if the discussion link above is just for contributors..if not would love to get access. Otherwise, if you need any early adopter validation we'd be happy to try it out.  Our primary use case for this would be API-driven, keeping user repository access in sync with another system we have that manages user entitlements.

**Comment by xaleeks** [comment]
@turnham Thanks, you can detail your requirements on this ticket here or email me.  We are looking to make a couple of changes, mainly 1) the ideas described here around grouping of users and permission mappings to existing projects as well as 2) add more granularity to existing harbor tenancy (ie think sub-project within a project) for better isolation.  I've shared the doc w/ you

**Comment by DASXCE** [comment]
Hi Team,

Are there any plans to allow Project Admins, maybe even Masters, to configure and trigger replications for their own projects?

**Comment by hyy0322** [comment]
@xaleeks Hi, I'v requested doc access. Please let me in, thanks.

**Comment by phin1x** [comment]
@xaleeks are there any updates on this issue or a timeline when this feature will be implemented? especially granting global roles to users would be great.

**Comment by manuelnucci** [comment]
@xaleeks what would be the best way with the current features of Harbor to have granularity by Users and Robot Account? One project per microservice? 

Does this issue aims to ease the management of different artifacts of different environments in the same project, and provide specific Service Accounts to specific pipelines?

**Comment by adamkdean** [comment]
This is a feature we would also find super useful. Currently we're restricted by the inability to have per-repo permissions for robot users.

**Comment by warmfusion** [comment]
We have the same problem.
We're using gitlab-ci to build projects, and have different project teams pushing containers to their respective projects in harbor.
Some teams collaborate and use eachothers images as part of their CI/CD pipelines but are unable to access the images due to the robot being used for their own project not being able to read from the other project.

If robots could be granted membership like regular users we'd be able to progress, however when attempting to add `robot@project+ci-bot` to the other project I recieve a server error.

**Comment by burghard-britzke-drv** [comment]
This issue is open since 2019 with no commit until now. Are there  any plans to work at a finer grained access control by the harbor project?

**Comment by adamkdean** [comment]
It's relevant, it's just that we're still waiting for the feature

**Comment by wirwolf** [comment]
Yes. I can not migrate from gitlab to harbor without this functionality

**Comment by wy65701436** [comment]
ping @qnetter for awareness

**Comment by nbatchelder** [comment]
I'm also very interested in this feature. Right now, we're forced to create one project per repository to achieve the flexibility we need.

**Comment by wirwolf** [comment]
Yea. I also wait this feature for migrate my projects in Harbor

**Comment by MaaxGr** [comment]
Hey @xaleeks,
is the discussion still in the google doc above?
https://docs.google.com/document/d/1AM9XEUjSepv7EgvjSpJXP-2pSBTrExoJsj8XKYN1sZ4/edit

Can you add access for google@maax.gr?

Kind Regards 
Max

**Comment by YohannHammad** [comment]
Hello! Do we have any idea of the progress?

**Comment by HectorB-2020** [comment]
To bump up this discussion :smiley_cat: I'm saying that this feature is more than welcomed from our side.

**Comment by nuxwin** [comment]
Feature really needed. Per project repository access management rather only per project access management.

**Comment by colbylarue** [comment]
We would really like to use this feature to have fine grained access controls on a per image basis that also takes into account classification labels on an image:tag version. We can make it work using the per repository access, but would also be nice to have classification labels.

**Comment by nishantmunjal7** [comment]
Is there any work being done on this ?

**Comment by gabrielcosi** [comment]
+1

**Comment by wirwolf** [comment]
cc. @stonezdj
